### PR TITLE
Minitest PoC

### PR DESCRIPTION
**Pull request by cristianbica**
This adds support for minitest. While it's working for me (using it on a rails 4 app and tested it on a rails 5 apps) it's still PoC (naming, tests missing). Minitest support has 2 parts:
- some rake tasks (in the `ruby` folder - `vscode:minitest:list` and `vscode:minitest:run`) which list or run tests - had to do this as there are a lot of differences between the tests runners in different rails versions 
- `minitestTests.ts` file which is copied from rspec and did the minimum changes that to make it work

This also adds an opt-in model for the ruby adapter. The user must add a workspace (or global) config to determine the testing framework used. 

TODOs / Ideas: 
- refactor the JS to have an abstract class which will be extended by minitest / rspec implementations. I'll probably work on that but don't have a lot of time and I'm not fluent in TS / JS
- I would remove some of the heavy lifting from JS (such as building the tree structure) and implement it in ruby
- move rspec list / runner into rake tasks

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "icon": "img/ruby-test-explorer.png",
   "author": "Connor Shea <connor.james.shea@gmail.com>",
   "publisher": "connorshea",
-  "version": "0.4.6",
+  "version": "0.5.0",
```
Comment (connorshea) [inline package.json]: Let's keep it at 0.4.6, I'll bump it when I release a new version
```suggestion
  "version": "0.4.6",
```

File: ruby/vscode/minitest/reporter.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        elsif r.passed?
+          base[:status] = "passed"
+        else
+          base[:status] = "failed"
+          base[:pending_message] = nil
+          e = r.failure.exception
+          base[:exception] = {
+            class: e.class.name,
+            message: e.message,
+            backtrace: Rails::BacktraceCleaner.new.clean(e.backtrace),
```
Comment (connorshea) [inline ruby/vscode/minitest/reporter.rb]: We should provide a fallback here, since we can't assume Rails is available.

**Comment by cristianbica** [comment]
> Thanks for the PoC :D
> 
> If you're fine with me pushing to your branch I can help fix this up a bit.

Sure. Go ahead.

> I'm somewhat wary of going too far with creating an abstract adapter class, but there are definitely a decent number of functions that can be made generic in the TS.

The idea is that over 80% of those classes are the same. So some DRY will make the code more readable. 
Another idea would be to have rake tasks for RSpec also so we could have only one ts file which runs commands based on a testing framework var.

**Comment by connorshea** [comment]
Okay, so I've abstracted the common functions from the two test framework files into an abstract Tests class, so now `rspecTests.ts` and `minitestTests.ts` are both around 200 lines vs 500-600 lines before. :D You were definitely right about the files needing to DRY'd up, I didn't think the files were quite so similar.

I've tested the functionality of both the Minitest and RSpec test runners and as far as I can tell they both work the same as they did before this refactor.

Two minor problem with the Minitest runner:

- It doesn't mark the test as running when it runs, it just changes from the Pending state to Success/Failure. Not a huge deal, but it'd be nice to have it fixed if possible.
- The 'rakeCommand' setting could probably be clearer about it being used for the Minitest runner. Maybe we should change it to `minitestCommand` instead?

The other thing I'm concerned about would be how we transition from the runner being RSpec-only to supporting Minitest. I'd rather not make the extension stop working for people without them being made aware of the change.

We could add a notification that will pop-up the first time the extension runs that says something about needing to select a test framework if 'none' is selected, and then a week or two later just remove that notification. Alternatively, we could just have it default everyone to RSpec? I don't think we'd ever be able to change the default to 'none' in that case, though. If the user doesn't explicitly choose an option in their settings file, the extension will just stick to whatever the default is.

**Comment by connorshea** [comment]
For what it's worth, the Minitest suite I tested on is a Rails 5.2 app ([My MDN Compat Data Explorer](https://github.com/connorshea/mdn-compat-data-explorer)) and it seems to work fine. What problems did you want to fix for Rails 5 apps?

**Comment by cristianbica** [comment]
> Okay, so I've abstracted the common functions from the two test framework files into an abstract Tests class, so now `rspecTests.ts` and `minitestTests.ts` are both around 200 lines vs 500-600 lines before. :D You were definitely right about the files needing to DRY'd up, I didn't think the files were quite so similar.

Looking very good. Forgot to tell you that I first commited to the repo the minitest TS file as an exact clone of rspec one and then did another commit with just the changes for minitest to see he difference (https://github.com/connorshea/vscode-ruby-test-adapter/pull/14/commits/d019436836ce8dede9cc63d51f8ee00b99dd63ae#diff-41ac763294d98cd851bb39864206da2cR2)

> I've tested the functionality of both the Minitest and RSpec test runners and as far as I can tell they both work the same as they did before this refactor.
> 
> Two minor problem with the Minitest runner:
> 
> * It doesn't mark the test as running when it runs, it just changes from the Pending state to Success/Failure. Not a huge deal, but it'd be nice to have it fixed if possible.

Will take a look at that. AFAIR minitest has a `prerun` callback for reporters so it should be fairly easy.

> * The 'rakeCommand' setting could probably be clearer about it being used for the Minitest runner. Maybe we should change it to `minitestCommand` instead?

Thought about that myself and had it like this initially but I'm very picky with names :) this is not actually a minitest command it's a rake command. What we really care about is if the user uses bundler or not (if we should run `bundle exec rake` or just `rake`). Maybe use a boolean `minitestUseBundler` and explain. What do you think?

> The other thing I'm concerned about would be how we transition from the runner being RSpec-only to supporting Minitest. I'd rather not make the extension stop working for people without them being made aware of the change.
>
> We could add a notification that will pop-up the first time the extension runs that says something about needing to select a test framework if 'none' is selected, and then a week or two later just remove that notification. Alternatively, we could just have it default everyone to RSpec? I don't think we'd ever be able to change the default to 'none' in that case, though. If the user doesn't explicitly choose an option in their settings file, the extension will just stick to whatever the default is.

Well we're doing a pretty big assumption here that people have RSpec. My main challenge is that whenever I opened a rails folder with minitest I would get an error about rspec command failed. Here is what I think we should do:
- Having a global value `minitest` or `rspec` is not the best for this project so in most cases people will have a project / workspace specific configuration
- Be proactive about detection: have an `auto` mode which is the default which will look for rspec in Gemfile and if it's there than we use rspec and else we'll use minitest. 
- Try to be more careful about when we start loading the test. Not sure when we start loading tests now and what are out current capabilities but assuming that "when people have the extension installed and open a folder which has some ruby files then they want to use the extension" it's pretty big. So I would also explore something in this area. Maybe start loading stuff when they open the Test Explorer panel.

**Comment by cristianbica** [comment]
> For what it's worth, the Minitest suite I tested on is a Rails 5.2 app ([My MDN Compat Data Explorer](https://github.com/connorshea/mdn-compat-data-explorer)) and it seems to work fine. What problems did you want to fix for Rails 5 apps?

Cool. Currently the `minitestFolder` configuration is not used all over the places (like the ruby files are assuming its' the `test` folder (which is 100% of all projects I've seen using minitest).

**Comment by cristianbica** [comment]
Added a commit (https://github.com/connorshea/vscode-ruby-test-adapter/pull/14/commits/670da682929955362e57eb27750dd88ce4cfbc92) to fix the status of the tests while running. The output to stdout was not synced so ruby would buffer to save i/o. Also added support to parse while tests are running the statuses of running and pending.

**Comment by cristianbica** [comment]
What's bugging me while using it is that if I add new tests I have to manually refresh. Do you have an example (of other language adapter) on how they are doing it?

**Comment by connorshea** [comment]
I assume the Python extension probably has it. Maybe Mocha as well. Unfortunately I don’t know exactly how they’ve implemented it.

That’s what #5 is for, we’d need to implement a test watcher that can reload the tests whenever a file in the test directory changes.

Unless the change is really minimal, I’d probably prefer to implement that in a separate PR after this one.

**Comment by connorshea** [comment]
It looks like the jasmine adapter implements it here with the onDidSaveTextDocument event: https://github.com/hbenl/vscode-jasmine-test-adapter/blob/master/src/adapter.ts

Seems like it shouldn’t be too bad to implement, I’d still like to have it as a separate PR though :) I might do it once we’re done with this.

**Comment by cristianbica** [comment]
Played a bit last night and a naive approach would be something like below. The below handles delete / renames also. Would have to check if the file changed is in the test folder. Also a nice one would be to be able to reload only tests from that file.

```diff
diff --git a/src/adapter.ts b/src/adapter.ts
index 8d87d33..96eeef0 100644
--- a/src/adapter.ts
+++ b/src/adapter.ts
@@ -27,6 +27,7 @@ export class RubyAdapter implements TestAdapter {
     this.disposables.push(this.testsEmitter);
     this.disposables.push(this.testStatesEmitter);
     this.disposables.push(this.autorunEmitter);
+    this.disposables.push(this.createWatcher());
   }
 
   async load(): Promise<void> {
@@ -85,4 +86,21 @@ export class RubyAdapter implements TestAdapter {
     let testingFramework: string = (vscode.workspace.getConfiguration('rubyTestExplorer', null).get('testingFramework') as string);
     return testingFramework || 'none';
   }
+
+  private createWatcher(): vscode.FileSystemWatcher {
+    let watcher = vscode.workspace.createFileSystemWatcher("**/*.rb");
+    watcher.onDidChange((file) => {
+      this.log.info(`Reloading test as a file changed: ${file.fsPath}`);
+      this.load();
+    });
+    watcher.onDidCreate((file) => {
+      this.log.info(`Reloading test as a file was created: ${file.fsPath}`);
+      this.load();
+    });
+    watcher.onDidDelete((file) => {
+      this.log.info(`Reloading test as a file was deleted: ${file.fsPath}`);
+      this.load();
+    });
+    return watcher;
+  }
 }
```

**Comment by cristianbica** [comment]
As for this PR I guess the only remaining item is related to configuration options. 
- use the test folder from config 
- make a decision on the minitestCommand / rakeCommand / use bundler

**Comment by connorshea** [comment]
Unfortunately I think reloading tests only from a given file is kind of a limitation of the main Test Explorer extension. We can kind of hack it in if we wanted by just taking the current test structure and  amending it with whatever data is returned for the changed file.

**Comment by connorshea** [comment]
I added the test watcher in this PR since it's not too much additional code, and I wanted to add it without causing conflicts in this PR, or needing to wait until after its been merged.

I used `onDidSaveTextDocument` since I think that covers all three cases (editing, creation, deletion) at once.

I also renamed `testingFramework` to `testFramework`.

**Comment by connorshea** [comment]
>Well we're doing a pretty big assumption here that people have RSpec. My main challenge is that whenever I opened a rails folder with minitest I would get an error about rspec command failed. Here is what I think we should do:
>    - Having a global value minitest or rspec is not the best for this project so in most cases people will have a project / workspace specific configuration
>    - Be proactive about detection: have an auto mode which is the default which will look for rspec in Gemfile and if it's there than we use rspec and else we'll use minitest.
>    - Try to be more careful about when we start loading the test. Not sure when we start loading tests now and what are out current capabilities but assuming that "when people have the extension installed and open a folder which has some ruby files then they want to use the extension" it's pretty big. So I would also explore something in this area. Maybe start loading stuff when they open the Test Explorer panel.

I quite like the idea of an `auto` option as the default test framework option, though I'm not sure how exactly to pull it off.

Parsing the Gemfile like you suggested would be possible, but I'm not sure if there'd be potential problems with that given how weird Gemfiles can be. We could also run `bundle list` and check it for `rspec-core`/`minitest`, which might be less likely to cause problems.

I agree with the problem about the the extension being over-eager when opening any projects with a Ruby file in it. My initial thinking was that it would be better to expose debug info by default than to silenty fail, but I think it'd mostly be an edge-case for the extension to fail to find tests, so we can be more conservative about when the extension puts out those notifications.

I think we should load the tests as soon as possible, regardless of whether the test panel has been opened. RSpec tests can take a while to load, so loading them in the background before the user wants to see them is generally better UX.

**Comment by connorshea** [comment]
The extension will now detect the test framework automatically, though you can also force it to use one or the other, or none :)

**Comment by connorshea** [comment]
TODO (pretty much the same as before 😄):

- Use configured `minitestFolder` in the minitest Rake task. I guess we can just pass the minitestFolder into the Rake task as an argument?
- Decide what to do about the `rakeCommand`. Personally, I still think it should be called `minitestCommand`. I understand the rationale for `rakeCommand`, but IMO it's secondary to the user being able to clearly tell what config option they want to set when using Minitest. We can explain in the description of the config setting that it needs to be a rake command, if you think that'd help.

I'll update the README and Changelog after the PR is merged, so you don't have to worry about that. Unless you have any other suggestions, I think those two are the only things blocking the merge. :)

**Comment by cristianbica** [comment]
I won’t be able to work on those this weekend but the changes are pretty simple


-- 
Cristian Bica

> On Jun 1, 2019, at 09:10, Connor Shea <notifications@github.com> wrote:
> 
> TODO (pretty much the same as before 😄):
> 
> Use configured minitestFolder in the minitest Rake task. I guess we can just pass the minitestFolder into the Rake task as an argument?
Probably as env var. the rake tasks will use that to add to ruby path and as default value in some places (instead of ./test/ now)
> Decide what to do about the rakeCommand. Personally, I still think it should be called minitestCommand. I understand the rationale for rakeCommand, but IMO it's secondary to the user being able to clearly tell what config option they want to set when using Minitest. We can explain in the description of the config setting that it needs to be a rake command, if you think that'd help.
Yep. I agree with that. Anyway the default bundle exe rake will be enough for the majority of the users

**Comment by connorshea** [comment]
Alright, I think it's good to merge now 🎉 

Thanks for the contribution! :D

**Comment by cristianbica** [comment]
Looking good man. Thanks

**Comment by cristianbica** [comment]
Next step I’ll be looking if we can parse ruby ast dump to figure out which are the test. Should be blazing fast so we can reload a file as you type

**Comment by connorshea** [comment]
Just released 0.5.0 👍 

Parsing the AST is something I've briefly considered, but haven't really gone through with just because I doubt it'll be easy to make it consistent with the output of RSpec/Minitest, but you're certainly welcome to try :)
